server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -414,8 +414,14 @@ PublishWithJoinResponse handlePublishRequest(PublishRequest publishRequest) {
         assert publishRequest.getAcceptedState().nodes().getLocalNode().equals(getLocalNode())
             : publishRequest.getAcceptedState().nodes().getLocalNode() + " != " + getLocalNode();
 
+        final ClusterState newClusterState = publishRequest.getAcceptedState();
+        if (newClusterState.nodes().isLocalNodeElectedMaster() == false) {
```

Comment:
I wasn't really able to benchmark whether this special handling of master is helpful or not admittedly (though I can't see how it would ever hurt). But given that this code is running on the single threaded coordinator pool, I found it easier to reason about master performance by leaving the special case in place here.